### ISSUE TITLE
Aggregate machine-readable provider network directories and drug formularies into tabular format

### ISSUE THREAD

**Issue by dportnoy** [demand-driven-open-data/ddod-intake#56]
### Background

In November 2015, the Centers for Medicare & Medicaid Services (CMS) enacted a new regulatory requirement for health insurers who list plans on insurance marketplaces. They must now publish a machine-readable version of their provider network directory and drug formulary, publish it to a specified JSON standard, and update it at least monthly.  This data has just recently become accessible to the public.  Some of its uses can be found in the [Bayes Impact hackathon "prompts"](http://bayeshack.org/health-and-human-services.html) or in [at least 7 DDOD use cases](http://hhs.ddod.us/wiki/Concept:_Group_use_cases_for_provider_registry).
### Challenge

While these newly available datasets can significantly benefit consumer health applications and be used in a range of healthcare analytics, the current format doesn't lend itself to doing so.  
- First, while there is a single "Machine-readable URL PUF" seed file, the actual data is scattered throughout thousands of URLs.
- Second, for analytics and aggregate operations, it would be beneficial to convert the JSON schema into a tabular data format.
### Request

Write code that does the following:
1. Crawls the URLs starting with the ["Machine-readable URL PUF"](http://download.cms.gov/marketplace-puf/2016/machine-readable-url-puf.zip) seed file found on this page: https://www.cms.gov/CCIIO/Resources/Data-Resources/marketplace-puf.html.
   -  (As of  3/1/2016, there are 636 plan URLs, 23,711 provider URLs and 1,662  formulary URLs.)
2. Converts the data from the JSON schema (https://github.com/CMSgov/QHP-provider-formulary-APIs) into a tabular format with the same fields.
   - (There may be some challenges converting multiple independent array fields from JSON to a tabular row.)
3. Aggregates the results into CSV or text delimited text files, such that we have files for: Plans, Providers, and Formularies.

Run the code and let us know where to find the resulting files.  We should be able to find a good home for them, so that they enjoy widespread use.

**Comment by dportnoy** [comment]
@marks, @ftrotter, @cornstein, @BAP-Jeff, @loranstefani, @Jordanrau, @HeatherBrotsos, @lloydbrodsky,
Since you're all past contributors, I wanted to see if any of you would take the challenge.  Feel free to pass it on to your colleagues as well.

**Comment by dportnoy** [comment]
In case it's helpful, I started working on the code, but just haven't had enough time to get to it lately.  Feel free to use.  You can find it here: https://github.com/dportnoy/health-insurance-marketplace-analytics

**Comment by marks** [comment]
@dportnoy - looks like you got pretty far.. would you mind documenting what it does so far and what areas need attention? Would definitely help so others dont accidentally reinvent the wheel

**Comment by dportnoy** [comment]
@marks You give me entirely too much credit.  But point taken.  I'll write up a summary and post it.

**Comment by dportnoy** [comment]
### Suggestions for JSON to tabular (CSV/TSV) column mapping

Translation from JSON schema into tabular should be straight forward...
- **Files:** Break into 3 core files by schema: plans, providers, drugs
- **Columns:** Schema fields get translated into individual columns
- **Arrays:** Biggest challenge is with arrays.  Multiple independent arrays are especially problematic, since you can't just create a new row with that value
  - **Option 1:** Just leave an array as a list within a single field.  
    - (Easiest to translate from JSON, but hard for analytics)
  - **Option 2:** Pivot array into separate columns.  
    - (Easiest for analytics, but there's a long tail of typos and outliers.)
  - **Option 3:** Normalize it by breaking arrays into separate files with foreign keys, such as plan id and provider NPI.  
    - (Great for loading into a relational database, but requires multiple steps before use for analytics.)

**Comment by marks** [comment]
I put something together really quickly that takes insurance provider's `index.json` URLs and flattens the data from the `formulary_urls` array.

Sample dataset: https://healthdata.demo.socrata.com/view/979j-m4qb
Code: https://github.com/marks/health-insurance-marketplace-analytics/commit/3507b4738fbb1ba75126e474bd6c24f0baf9d133

**Comment by cornstein** [comment]
Really interesting output, but I wonder if there's a way to make this a bit
more useful. Perhaps we can have a brief chat?

**Comment by marks** [comment]
@cornstein - definitely... That's just a quick visualization of the data. You can export it (all or a slice) or use the API as well, of course. 

Kind of hectic day but happy to jump on a call tomorrow or continue to chat here if you're interested in a specific view 

## 

Mark Silverberg
(m) 512 826 7004
http://twitter.com/skram

**Comment by dportnoy** [comment]
Here are some options for the end product, along with their uses.  They are listed in **increasing order of difficulty**:
1. AGGREGATE: Aggregate the crawled URLs into a small number of files, keeping the original [QHP schema](https://github.com/CMSgov/QHP-provider-formulary-APIs) format in JSON: plans, providers, drugs.
   - Eliminates the need to crawl thousands of URLs is no longer needed. 
   - Expect to run into file size challenges, since there are thousands of URLs and some files may be hundreds of megabytes.
   - This format may be adequate for loading into a NoSQL/document database for use with non-analytical consumer web apps.
2. FLATTEN: Make the format tabular, in the easiest way possible.  
   - Method: The files produced can line up exactly with the QHP schema entities pulled from URL links: plans, providers, drugs.  For arrays, you could put multiple values in a single field.  
   - The goal here is to create files that can be hosted ASAP and used for rudimentary analytics.
3. NORMALIZE: Normalize the work product from item # 2 into relational tables, tied with foreign keys.
   - This format is useful for loading into a relational database for use in consumer apps and web services, especially if they require calculations and transformations.  Convenient for codathons and challenges.
4. JOIN & PIVOT: Make a tabular format that’s easiest for analytics by creating one row per entity with all possible attributes.  
   - For example, in addition to the fields in the drugs.json schema and its plans subtype, include fields from  ["Machine-readable URL PUF"](http://download.cms.gov/marketplace-puf/2016/machine-readable-url-puf.zip) XLSX seed file: Marketplace Category, State, Issuer ID, Issuer Name.  You can find other joins between the entities of the QHP schema.
   - Expect to run into challenges for entities with multiple arrays that are free-form or have a long tail of values.

**Comment by marks** [comment]
I've made a bit more progress. The latest code is available at https://github.com/marks/health-insurance-marketplace-analytics/blob/master/flattener/flatten_from_index.py

I have code that _starts*_ to flatten all three file types: `drugs.json`, `plans.json`, and `providers.json`. I have only run `providers.json` and `drugs.json` on a few `index.json`'s as it can take a while to process. 

The flattened data can be explored/downloaded/called-with-an-API from the following links:

| File Type Flattened | Data Explorer <br /> (you can export from here too) | Data Grid <br /> (you can do roll ups and filtering here) | API Docs |
| --- | --- | --- | --- |
| `plans.json` | https://healthdata.demo.socrata.com/view/yjgn-upfm | https://healthdata.demo.socrata.com/d/x6bt-9iym | https://dev.socrata.com/foundry/healthdata.demo.socrata.com/55ch-w329 |
| `providers.json` | https://healthdata.demo.socrata.com/view/pqgg-pgsy | https://healthdata.demo.socrata.com/d/axbq-xnwy | https://dev.socrata.com/foundry/healthdata.demo.socrata.com/5ht2-73gj |
| `drugs.json` | https://healthdata.demo.socrata.com/view/979j-m4qb | https://healthdata.demo.socrata.com/d/jaa8-k3k2 | https://dev.socrata.com/foundry/healthdata.demo.socrata.com/6ngr-bz3d |

`*` Not all data is flattened at this time to decrease the number of rows. For example, `languages` in the provider file are serialized as a JSON array as are `addresses` and `specialties`

**Comment by dportnoy** [comment]
@BAP-Jeff, hi again.  Welcome to use case #56!  To answer your question from issue #52...  Having a ready to go dataset would be ideal for the Bayes Hack codathon for next weekend.  It's identified as the central dataset for building consumer healthcare apps.  You'll see a comment above that identifies 4 possible file types: https://github.com/demand-driven-open-data/ddod-intake/issues/56#issuecomment-210875780.  It's good to have a few people working on this, because some of the aspects (such as file sizes and cleanup for analytics) are challenging.  Having the code would be a useful resource for the community to refresh the data in the future.

**Comment by ghost** [comment]
Hello All,

As David knows, I have been playing with the drugs.json set for a while as
I build some commercial applications for the formulary space.  There are a
ton of challenges in using the data as supplied but overall it is a good
first year effort by the agency.

If I understand the challenges David laid out, I have been pursuing Option

# 4 for my app.  I am using a couple simple Python scripts to (1) read a

list of URLs from the Machine Readable PUF, then retrieve the drugs.json
files for each (2) then I run a second script to rip apart the json
documents into a flat pipe delimited file.

I am happy to share the code and/or the end result (warning around 35 to 40
million records).  What do we want for the hack-a-thon?  Have to admit I
have not participated in one before, so I am not sure if people are only
interested in the end result (upload the dataset to say Socrata or kaggle)
or do they want to understand the process, so they can roll their own?

Let me know which way the group wants to go.

Jeff

**Comment by marks** [comment]
@BAP-Jeff - awesome. I think both the data and the provenance (how you got it, including any scripts/documentation) would be much appreciated by participants as well as the community in general. Would be happy to look into hosting it onto Socrata if you can provide the raw data

Mark

**Comment by ghost** [comment]
Okay thanks.  I will put together some things tonight.  Much of my time in
working with this dataset is dealing with the data discrepancies (i.e.
multiple statuses on a drug for a plan, use of NDCs vs RxCUIs, missing
restrictions, bogus URLs), but showing off the millions of records that are
produced will be cool.  Plus hopefully making it more accessible will get
more eyes on it...and we can get those things corrected.

Here is the "process" I would propose that we show off (for the drugs.json
branch):
- Download latest Readable PUF
- Extract the details for the URLs and Plans, excluding dental and NOT SUBMITTED
  records
- Iterate over that list, pulling down the valid json files
- Then iterate over the json files and create csv files
- Then push them to Socrata and/or other spots

David, do you need a "deck" or some sort of visual to present?  Help me out
if I am going astray...

Jeff

**Comment by marks** [comment]
I dont want to make more work for you, @BAP-Jeff, but I'd be very interested in what your CSVs are looking like. I am running some code on the `index.json`s I found in a previously shared `xlsx` to flatten all three file types (but they may not have all important info... I am not a SME in this area

**Comment by ghost** [comment]
I actually use pipe delimiting but can easily change it out.  I am attaching an example source json file and what I am outputting (tried to find a small one).  It is very close to what you had shown off before Mark.  I can easily add some more columns to this from the Plan Attributes PUF file to be more complete (IssuerID, Region, etc), but sounded like you wanted to see what the current state was.  I also included the python script I used to parse the file.

The format of the output is:
- Source JSON file name
- PlanIDType
- PlanID
- DrugStatus
- DrugID
- PA
- ST
- QL

[Archive.zip](https://github.com/demand-driven-open-data/ddod-intake/files/222686/Archive.zip)

[HIXFormExtract.py.zip](https://github.com/demand-driven-open-data/ddod-intake/files/222687/HIXFormExtract.py.zip)

**Comment by marks** [comment]
@BAP-Jeff - Yup! Looks quite similar. I would think it would be very helpful to folks to include info from the Plan Attributed PUF. Will standby for files so that we can get them on Socrata\* (and Kaggle and other platforms) for dissemination/exploration/API acces

`*` Full disclosure: I work for Socrata and we're more than happy to help folks download the data in a number of formats. Can accept pipe delimited, no big deal.

**Comment by ghost** [comment]
Okay I will be gone most of the day but will get on this later today.

**Comment by dportnoy** [comment]
@BAP-Jeff, great to have you contribute!  Let's see if I could answer most of your questions...
- STORAGE: For now, the datasets generated can be stored wherever is most convenient for your methods.  They can always be copied elsewhere later.
  - Ultimately, Kaggle agreed to put the datasets in their HHS section for use with the codathon and other future activity: https://www.kaggle.com/hhsgov/health-insurance-marketplace
  - If needed, I can provide you with an S3 bucket to dump it in.  (But I still need to determine the most appropriate permissions and access mechanism.)
  - If Socrata is willing to host any portion of it, that's great too.
- FILE SIZES: You asked about potential sizes.  I think we need two sets of files: 
  - The **complete files** (with millions of records), which are obviously needed for analytics.  We may end up needing to break these into multiple files for manageability. 
  - A small **data sample** subset, which can be loaded for proof of concepts with applications.  Perhaps pick a few regions, but make sure the files are complete for a particular plan, so that joins are possible.
- EXAMPLES: For your pipe-delimited file, we'd need to be sure to include either a header record or package a dictionary with it.  For data.json processing, we can add more useful fields by joining with the ["Machine-readable URL PUF"](http://download.cms.gov/marketplace-puf/2016/machine-readable-url-puf.zip) XLSX file and possibly attributes from other entities (plans and providers).
- DECK: Yes, if you can put together a deck, it will be useful.  I think data visual slides would be compelling -- both for confirming assumptions and demonstrating surprising "aha" moments.

**Comment by ghost** [comment]
David,

I think I can take on all the above but may need a little help with the Deck.  Do you have a template you want to use - Powerpoint, Google, ??

I can produce the files tonight/tomorrow and upload them probably tomorrow.

@dportnoy give me the details on the kaggle stuff
@marks let me know where to drop the files for Socrata

David, did you want to reserve any slides for ..."Issues we found in the data" or does that diminish the good vibes?

Jeff

**Comment by marks** [comment]
Jeff- if I can just get delimited files, I can take care of loading it for ya. Sounds like kaggle is top priority for this event though ;)

## 

Mark Silverberg
(m) 512 826 7004
http://twitter.com/skram

**Comment by ghost** [comment]
Okay I will loop back around to you as time is short.

**Comment by dportnoy** [comment]
@BAP-Jeff  Google slides.  No specific template.  Just the white default.  

For Kaggle, upload to it will be done manually by their staff (since they're still working on making that functionality self-serve).  So we need a staging location before the data gets copied there.

**Comment by ghost** [comment]
Okay I have the file ready to go.  It is about 5GBs uncompressed (31 million records).  The file has the following table def:

create table Hackathon
(StateCode          varchar(2),
IssuerID            char(5),
StandardComponentID varchar(14),
MarketCoverage      varchar(18),
PlanMarketingName   varchar(142),
HIOSProductID       varchar(10),
ServiceAreaID       varchar(6),
FormularyID         varchar(6),
PlanType            varchar(9),
MetalLevel          varchar(12),
DrugStatus          varchar(100),
RxCUI               varchar(15),
DrugName            varchar(2000),
PAStatus            char(1),
STStatus            char(1),
QLStatus            char(1))

I pulled from Plan Attributes what I thought was the key information.  Note: PlanID in the Formulary Files is not the same as PlanID in Plan Attributes.  The analogous column is StandardComponentID, furthermore I ignored the CSR variations in Plan Attributes and joined the distinct records from Plan Attributes to the Formulary Files.  If this didn't make sense...well I guess trust me...

I have added a header record to the file and am now waiting for the zipping to finish up.  I assume it will get down to the 2GB range.  I can throw it up into a S3 bucket and share it I guess, or GDrive.  Let me know.  Could be a few days on the slides....

Jeff

**Comment by dportnoy** [comment]
@BAP-Jeff, definitely throw it up on either S3 or gDrive, whichever works best!  Ideally, you could also provide a smaller file that has a more manageable sample subset of this data, so that it can be used to test code.

**Comment by dportnoy** [comment]
@ftrotter, wanted to ping you and see if you're still working on any aspect of this.

**Comment by ghost** [comment]
Actually it zipped down to 250 megs so not such a monster.  I also did a sample file for all of Humana in Georgia.

**Comment by marks** [comment]
@BAP-Jeff - I think you missed a pipe in the header row

**Comment by ghost** [comment]
Thank goodness someone is QAing this!  Thanks @marks.  Be right back.

**Comment by ghost** [comment]
DRUGS file sample: https://drive.google.com/file/d/0B9yZheZrBn54RHk1SmszenNnYTg/
DRUGS file complete: https://drive.google.com/file/d/0B9yZheZrBn54MGplUW5BeVRhMDQ/

**Comment by dportnoy** [comment]
@marks apologies for the delayed reply on your questions in https://github.com/demand-driven-open-data/ddod-intake/issues/56#issuecomment-210890927.  

Unfortunately, I can't get your links to work.  I'm getting a message "Sign in with your Socrata ID".  But when I register a new ID, I get "The current user doesn't have access to this view".  For the API links, I'm getting a "Private Dataset" message.

**Comment by dportnoy** [comment]
@BAP-Jeff, it rocks!

DRUGS file complete [Hackathon.zip](https://drive.google.com/file/d/0B9yZheZrBn54MGplUW5BeVRhMDQ/) unzips (using `tar xzvf Hackathon.zip`):
- Size: 5.1GB
- Rows: 330,150,734 (including header row)
- Included header: StateCode|IssuerID|StandardComponentID|MarketCoverage|PlanMarketingName|HIOSProductID|ServiceAreaID|FormularyID|PlanType|MetalLevel|DrugStatus|RxCUI|DrugName|PAStatus|STStatus|QLStatus
- Alternate [schema-conforming](https://github.com/CMSgov/QHP-provider-formulary-APIs) header: state_code|issuer_id|plan_id|market_coverage|plan_marketing_name|hios_product_id|service_area_id|formulary_id|plan_type|metal_level|drug_tier|rxnorm_id|drug_name|prior_authorization|step_therapy|quantity_limit

DRUGS file sample [HackathonSample.zip](https://drive.google.com/file/d/0B9yZheZrBn54RHk1SmszenNnYTg/) unzips (using `tar xzvf HackathonSample.zip`):
- Size 22MB
- Rows 127,443 (including header row)
- Headers:  Same as complete file 

I verified the columns for the sample file align and match the header you provided.  So we might be good to go with the drug formulary files!

**Comment by marks** [comment]
@dportnoy mind testing those links again?

**Comment by dportnoy** [comment]
@BAP-Jeff, I believe you said you were also looking at **PLANS** files, right?

Some questions and possible improvements for **DRUGS**:
- The `ServiceAreaID` and `FormularyID` fields are equal.  Where are they coming from?
- Would it be possible to use _snake case_ and naming that's consistent with the [schema spec](https://github.com/CMSgov/QHP-provider-formulary-APIs), rather than _camel case_ and descriptive naming.  For example...
  - `plan_id`, instead of `StandardComponentID`  (Also add column `plan_id_type`, just in case, although it's typically HIOS_PLAN_ID)
  - `marketing_name`, instead of `PlanMarketingName`
  - `drug_tier`, instead of `DrugStatus`
  - `rxnorm_id`, instead of `RxCUI`
  - `prior_authorization`, instead of `PAStatus`  (use true/false, instead of y/n)
  - `step_therapy`, instead of `STStatus`  (use true/false, instead of y/n)
  - `quantity_limit`, instead of `QLStatus`  (use true/false, instead of y/n)
  - etc.
- Would you want to add a couple more fields from the PUF file:
  - `marketplace_category` (FFM/SPM)
  - `issuer_name` (in addition to the PlanMarketingName name you already have)

So exciting!

**Comment by dportnoy** [comment]
@marks  Progress!  Data grid and APIs seem to be working, but data explorer redirects to https://healthdata.demo.socrata.com/.  I can dig into the detail by tomorrow morning.

**Comment by marks** [comment]
@dportnoy think I fixed my typo. Gonna load @BAP-Jeff's file so that I can play with some aggregations etc via API as opposed to having to work with the full file which is a bit large ;)

**Comment by ghost** [comment]
@dportnoy, I have not been working with the PLANS files yet.

Regarding your questions:
- I just double checked ServiceAreaID and FormularyID.  I don't see them as
  being equal.  The third character is an "S" for ServiceArea or "F" for
  Formulary.  These items comes from the Plan Attributes PUF
- I will change the naming.  I was trying to be consistent with the actual
  Data Dictionaries on the PUF site
- Which PUF file has the marketplace_category and issuer_name?

**Comment by ghost** [comment]
Here is what the modified header would look like:

state_code|issuer_id|plan_id|market_coverage|plan_marketing_name|hios_product_id|service_area_id|formulary_id|plan_type|metal_level|drug_tier|rxnorm_id|drug_name|prior_authorization|step_therapy|quantity_limit

Sorry those pipes are hard to see.

**Comment by ghost** [comment]
@dportnoy, I am taking a quick look at plans.json.  I just downloaded them all and now playing with the extraction.  Probably not a big challenge.

**Comment by marks** [comment]
@BAP-Jeff's drugs.json extract is on Socrata exploration experience which also links to export and API at https://healthdata.demo.socrata.com/view/7unb-qpyc if anyone wants to use it to take a look. If there is a data dictionary explaining each field, I can put that in to provide a little more context to users. Hope it's helpful!

Mark

**Comment by dportnoy** [comment]
@marks, @BAP-Jeff, way to team up and get the DRUGS file built and hosted!

@BAP-Jeff, thanks for the updated header record!  To answer your question, `marketplace_category` and `issuer_name` can be found in the ["Machine-readable URL PUF"](http://download.cms.gov/marketplace-puf/2016/machine-readable-url-puf.zip) seed file.

**Comment by ghost** [comment]
@dportnoy, I see the issuer_name but I don't see the marketplace_category.

**Comment by dportnoy** [comment]
@marks, I was finally able to get to all the links in https://github.com/demand-driven-open-data/ddod-intake/issues/56#issuecomment-210890927.  Great start on those files!  I see that you combined arrays and hierarchies into single fields for simplicity.  I think next step should be to generate the complete files, so that at least we have something useful as a starting point for the weekend.  We could always improve later as time permits.  As with @BAP-Jeff's DRUGS file, we can use the datasets you already built as the sample subsets for app development and testing.  

Is there a naming convention for fields that start with underscore or contain period?

Not crucial but...  Consider pulling in a few of the attributes from the ["Machine-readable URL PUF"](http://download.cms.gov/marketplace-puf/2016/machine-readable-url-puf.zip) seed file or other cross-reference between PLANS-PROVIDERS-DRUGS that might be useful for analytics.

**Comment by ghost** [comment]
@dportnoy, I have the plans.json material parsed.  What is less obvious is how we want to conflate these guys into a single record...actually I don't think it makes sense to do so. Have you looked at it at all?  I could create  plan, network, formulary and cost_sharing files if you like, joined by plan_id or something.  Looks like the benefit piece is a work in progress.

Let me know. Unlike the formulary these files are extremely easy to work with (read: small), so I can iterate much more quickly....

**Comment by dportnoy** [comment]
@BAP-Jeff, on your questions:
- https://github.com/demand-driven-open-data/ddod-intake/issues/56#issuecomment-211927446: Looks like CMS took out the `marketplace_category` field from the latest PUF.  Here's a PUF from December 2015 that has it: https://github.com/demand-driven-open-data/ddod-intake/blob/master/_files/Machine_Readable_PUF_2015-12-21.xlsx
- https://github.com/demand-driven-open-data/ddod-intake/issues/56#issuecomment-211959206: Yes, it's not as obvious what to do for flattening out PLANS and PROVIDERS.  That's what prompted the discussion of the 4 types of approaches in https://github.com/demand-driven-open-data/ddod-intake/issues/56#issuecomment-210875780.  You're suggesting a normalized approach (# 3), which is fine.  In general, let's get the complete data loaded first in whatever format is the easiest.  Then we can work on optimizing it for analytics or applications, as time allows.  Sound good?

**Comment by dportnoy** [comment]
@BAP-Jeff, @marks, @cornstein, if time allows it might be useful to figure out how to represent the plans that have **missing data**.  In other words, those who either:
1. Didn't provide an index.json URL at all
2. Provided a URL, but it didn't work
3. Had a working URL, but the content didn't conform to the schema

Seems that having missing values would be useful for both analytics and consumer applications.  The question is where to best capture this info.  Thoughts?

**Comment by ghost** [comment]
@dportnoy, I am out for the rest of the week EOB Wednesday.  I am going to prioritize doing another cut of the formulary data with the additional fields.  Then I will create the plans data.  Unlikely I am going to get to the slides at this moment.  Sorry.

As for your piece on who's missing, #1 is easy, #2 takes a bit of manual entry, and #3 (at least for me ends up in bucket #2).

Should I just add a column to say the machine readable puf with "GOOD", "BAD" for the URL provided?

**Comment by ghost** [comment]
All, here is my quick attempt to stitch something together for plans.json.

Four files joined on plan_id and/or plan_id-drug_tier.  Take a look, just one plan for now.

[plans.zip](https://github.com/demand-driven-open-data/ddod-intake/files/226562/plans.zip)

**Comment by dportnoy** [comment]
@BAP-Jeff,

> Should I just add a column to say the machine readable puf with "GOOD", "BAD" for the URL provided?

Sure that could work.  (If it's easier, we can have a separate addendum file with the same header, but containing just the plans that couldn't be processed.  Analysts could simply append this file to the main one, if they want to include the missing data points.)

**Comment by dportnoy** [comment]
@BAP-Jeff, on https://github.com/demand-driven-open-data/ddod-intake/issues/56#issuecomment-212055405

> ==> PlanOut.pip <==
> plan_id|marketing_name|summary_url|marketing_url|formulary_url|plan_contact|last_updated_on
> 
> ==> PlanNetworkOut.pip <==
> plan_id|network
> 
> ==> formularyOut.pip <==
> plan_id|drug_tier|mail_order
> 
> ==> costOut.pip <==
> plan_id|drug_tier|pharmacy_type|copay_amount|copay_opt|coinsurance_rate|coninsurance_opt

Looks good!  Let's load.  A couple observations for future improvement:
- Should the field in PlanNetworkOut.pip be `network_tier`?
- Consider including `telemedicine` field

**Comment by ghost** [comment]
Sure I will update the network_tier. As for telemedicine. Only one plan provided it (all false). I can create a benefit file but it would be short and boring....

**Comment by dportnoy** [comment]
@BAP-Jeff, Sounds good.  No need to create telemedicine if there's no data.

Since there's not time for slides...  From your prior analysis of the data, do you recall any observations that might be interesting to discuss?

**Comment by ghost** [comment]
@dportnoy, I am finishing up the plan stuff and unfortunately a few big issuers supplied malformed json documents...I am guessing they rely on the same vendor to produce these so that is why I see the same mistake a few different places.

In a nutshell:

"network": [ {"network_tier": "STATEWIDE" }], "formulary":  {"drug_tier": "HIGHEST-COST-SHARE", "mail_order": true, "cost_sharing": [ {"pharmacy_type": "1-MONTH-IN-MAIL", "copay_amount": 0.00, "copay_opt": "NO-CHARGE-AFTER-DEDUCTIBLE",.........

They are missing the [ bracket after formulary.  Oh well.

So first observation is validate your json files!  I'll come back to you with some observations, but let me get the files squared away first...

**Comment by ghost** [comment]
Here is the full set of plans.json files.

[plans.zip](https://github.com/demand-driven-open-data/ddod-intake/files/226850/plans.zip)

I'll get the formulary files up later together along with observations.

**Comment by ghost** [comment]
@dportnoy I "hearted" a comment from earlier today.  Can you take a look to see if you still see an issue?  I am holding on producing the final files until I hear.

I will add in issuer and marketplace from the old PUF file at that time too.

**Comment by dportnoy** [comment]
@BAP-Jeff, Nice job on [plans.zip](https://github.com/demand-driven-open-data/ddod-intake/files/226850/plans.zip)!  These files are pleasantly small!  We don't even need separate sample files.  I'm assuming they will be eclipsed by PROVIDERS files.

> **Row counts**
>    22,295  cost.pip
>    36,226  formulary.pip
>    13,969  network.pip
>    12,233  plan.pip

Some unexpected line breaks in plan.pip might cause problems with loading.  For example:
- 31609PA0150004
- 16204OH0020010

**Comment by dportnoy** [comment]
@BAP-Jeff,

> I "hearted" a comment from earlier today. Can you take a look to see if you still see an issue? I am holding on producing the final files until I hear.

Is it this one: https://github.com/demand-driven-open-data/ddod-intake/issues/56#issuecomment-211861941

> Regarding your questions:
> - I just double checked ServiceAreaID and FormularyID.  I don't see them as
>   being equal.  The third character is an "S" for ServiceArea or "F" for
>   Formulary.  These items comes from the Plan Attributes PUF

**A:** Sure.  Keep both.

> - I will change the naming.  I was trying to be consistent with the actual
>   Data Dictionaries on the PUF site

**A:** Already done.

> - Which PUF file has the marketplace_category and issuer_name?

**A:** See https://github.com/demand-driven-open-data/ddod-intake/issues/56#issuecomment-211976114 for answer

**Comment by ghost** [comment]
Uh oh let me look. 

Is someone getting the physician file done for you?

**Comment by ghost** [comment]
Regarding line breaks...oh fun check out the source:

"plan_id_type" : "HIOS-PLAN-ID",
  "plan_id" : "31609PA0150004",
  "marketing_name" : "Personal Choice PPO Gold Classic\n $1,000 $15/$30/80%",
  "summary_url" : "https://www.ibx.com/ffm/shop2016",

Nice little \n in the marketing name

Here is the corrected version 
[plans.zip](https://github.com/demand-driven-open-data/ddod-intake/files/227085/plans.zip)

**Comment by ghost** [comment]
@dportnoy still working on the formulary files to add in issuer and marketplace.  I need to kick off a new join and then download, roundtrip is a few hours.

Will late in the day work tomorrow or even Thursday?  This daytime job is getting in the way.

**Comment by dportnoy** [comment]
@BAP-Jeff, yes, funny about "\n".  Sneaky.  Tomorrow is OK.  We'll let that day job thing slide this time.  :-P

@marks, @ftrotter, want to tackle PROVIDERS?  @marks, was my feedback on https://github.com/demand-driven-open-data/ddod-intake/issues/56#issuecomment-211932545 sufficient?  @ftrotter, if I recall correctly all things provider networks are your specialty!

**Comment by ghost** [comment]
@dportnoy, hey I have added in marketplace and issuer name to the file, but we lost around 500K records (some issuer wasn't around in December I guess).  Whatcha think?  Use this new file or leave the old one?

**Comment by ghost** [comment]
Or we could just leave those columns blank...

**Comment by dportnoy** [comment]
@BAP-Jeff, do an outer join, so that you don't lose any records.  For missing values, perhaps leave blank.  It can be figured out and backfilled later.

And in case this is helpful...

> -   FFM vs SPMs for plan year 2016:
>   -   19 FFM states: Alabama, Alaska, Arizona, Florida, Georgia, Indiana, Louisiana, Mississippi (individual market), Missouri, New Jersey, North Carolina, North Dakota, Oklahoma, Pennsylvania, South Carolina, Tennessee, Texas, Wisconsin, Wyoming
>   -   15 SPM states: Arkansas (individual market), Delaware, Iowa, Illinois, Kansas, Maine, Michigan, Montana, Nebraska, New Hampshire, Ohio, South Dakota, Utah (individual market), Virginia, West Virginia

**Comment by ghost** [comment]
Here ar the updated formulary files....first one is the sample

[https://drive.google.com/file/d/0B9yZheZrBn54Y3ZEME5LcnA0S0E/view?usp=sharing](url)

[https://drive.google.com/file/d/0B9yZheZrBn54M2RuTW1xU2RWOXc/view?usp=sharing](url)

**Comment by dportnoy** [comment]
@BAP-Jeff, great!  Remind me the differences from the prior ones please.

**Comment by ghost** [comment]
Added in Issuer name and marketplace

**Comment by marks** [comment]
I'll take the new Formulary file and upload to Socrata in case anyone wants to work off of that as opposed to having to dealing with the 6+GB file :)

**Comment by dportnoy** [comment]
Great, @marks.  I appreciate it.  I'm working on a page with all the links.

That said, we still need PROVIDERS files.

**Comment by ghost** [comment]
I could maybe get to it tomorrow morning if you are desperate.

**Comment by dportnoy** [comment]
That would be beyond fantastic!

**Comment by ghost** [comment]
I just kicked off a process to download all the provider.json files.  I have gotten through the first 10, initial impression is that these files are MASSIVE.

This could take a while to process.

**Comment by marks** [comment]
@BAP-Jeff - that's why I initially focused on formulary ;) - Anyway I can help by running some scripts locally for you and/or on some cloud servers?

**Comment by ghost** [comment]
@marks, ha!  Thanks for the offer.  I can grab the files, but based on some back of the envelope calculations this is going to be multi-terabytes of source data.

I'll check in later.

**Comment by dportnoy** [comment]
@BAP-Jeff, @marks, I was afraid of that.  A few options:
1. **Break up** into multiple files
2. **Normalize** the heck out of the data  
   - It should help that multiple plans use identical networks and providers.json URLs.  Perhaps use a separate "issuer_provider" file that joins `issuer_id` to `providers_url`.  Then the "providers" file doesn't need to repeat entries that come from the same `providers_url`.
   - But this approach is a pain for analytics purposes.
3. **Don't load all fields**, focusing just on the ones not found anywhere else
   - While this doesn't change the cardinality and therefore the number of rows, it can reduce the file sizes significantly. Perhaps include key fields:  npi, type, state, zip, speciality, accepting, plan_id, network_tier.
   - This approach may be sufficient for many analytics purposes.  Unfortunately, it may not be as useful for consumer applications, which could benefit from fields such as address and language

---

None of these are ideal.  Perhaps to get something going, we can focus on loading a subset of the data.  So pick either one interesting large state (like TX or FL) or small state (like AK or SD) to load.  This would at least allow for use with consumer apps and certain types of analytics.

**Comment by marks** [comment]
@BAP-Jeff, @dportnoy - the latest (https://github.com/demand-driven-open-data/ddod-intake/issues/56#issuecomment-212898336) formulary files are API enabled at the following link. Let me know how else I can help, of course.

https://healthdata.demo.socrata.com/view/xc22-8t66

**Comment by dportnoy** [comment]
Thanks! That leaves just the providers files.

**Comment by ghost** [comment]
I downloaded all the provider json files last night.  Looks like maybe the "big" ones were at the top of the list and they got better after that.  As we discussed/suggested, I am breaking them into a number of tables - indiv_languages, fac_addresses, etc...  Still going to be big files, but seems like things are running....

**Comment by ghost** [comment]
FYI, we have some healthy sized files.  I just exhausted the memory on my box.  It will take me a few to spin up a big honker on AWS to run these big guys...

Maybe I should just give up on these for now and deal with them later.  @dportnoy what is the timing on all this?  Today right?

-rw-rw-r-- 1 ubuntu ubuntu 2267461069 Apr 21 21:11 PROVJSON3_201604212013.json
-rw-rw-r-- 1 ubuntu ubuntu 2170298888 Apr 21 22:29 PROVJSON2903_201604212013.json
-rw-rw-r-- 1 ubuntu ubuntu 1544176291 Apr 21 22:18 PROVJSON2891_201604212013.json
-rw-rw-r-- 1 ubuntu ubuntu 1456651939 Apr 21 23:08 PROVJSON2915_201604212013.json
-rw-rw-r-- 1 ubuntu ubuntu 1081970470 Apr 21 22:44 PROVJSON2910_201604212013.json
-rw-rw-r-- 1 ubuntu ubuntu 1073854978 Apr 22 01:46 PROVJSON3425_201604212013.json
-rw-rw-r-- 1 ubuntu ubuntu 1029692658 Apr 22 00:21 PROVJSON2987_201604212013.json
-rw-rw-r-- 1 ubuntu ubuntu 1014403038 Apr 21 23:15 PROVJSON2916_201604212013.json
-rw-rw-r-- 1 ubuntu ubuntu  976208707 Apr 21 22:36 PROVJSON2909_201604212013.json
-rw-rw-r-- 1 ubuntu ubuntu  965357940 Apr 21 21:58 PROVJSON2833_201604212013.json

**Comment by dportnoy** [comment]
@BAP-Jeff, for now could you create a sample file and pick one full state to crawl.  (See bottom of my note above.)  Since the size would be manageable, you could create the simplest layout possible.

**Comment by ghost** [comment]
@dportnoy actually that is tougher than it sounds.  To get a full state would imply that we would need to load all the files then query the state we want out of it.  I can work towards that but it will take me some time to get a machine that can parse the 2GB files.  What is the timing?

Here is what the header files will look like for eight files (anything obvious I am missing?):

# Plan Table - for Indiv and Facilities (I guess I could create separate ones...)

plan_id_type|plan_id|network_tier|npi

# Individual Files

npi|prefix|first|middle|last|suffix|accepting|gender|last_updated_on

npi|language

npi|specialty

npi|address|address_2|city|state|zip|phone (I could combine with facility address file...)

# Facility Files

npi|facility_name|last_updated_on

npi|facility_type

npi|address|address_2|city|state|zip|phone

**Comment by dportnoy** [comment]
@BAP-Jeff, 

**One-state subset:** I could be wrong, but I thought you start with the `state` associated with each issuer in the  [Machine Readable PUF 02-22-2016.txt](https://github.com/demand-driven-open-data/ddod-intake/files/231940/Machine.Readable.PUF.02-22-2016.txt).  Not sure if each shows nationwide networks, but I'm sure some have narrow networks that wouldn't be as big.  If that logic is incorrect or doesn't work well from a workload perspective, please do what makes most sense.

**Sample file:**  Besides the one-state option, it would be helpful to have a small sample file to work with -- regardless what data is in it.

**Timing:** Ideally by end of day.  But I'll start writing up links to the files we already have in parallel.  You've done so much already.  I really appreciate it!

Looking at fields next...

**Comment by ghost** [comment]
@dportnoy, I think what you are saying is more or less right.  What I am
seeing is that an Issuer can be operating across multiple states, so the
url listed for say Aetna/Texas is the same url as Aetna/Florida.  I could
(as I think you suggest), go after the urls listed for a state and if there
is spill over to other states so be it.  But looks like national plans are
just that....they show up everywhere which makes the task tough.  I am
making progress on loading everything though.

Do you have a state that you would like me to grab?

**Comment by dportnoy** [comment]
@BAP-Jeff, on fields...

I think we need to add `plan_id` and `network_tier` to INDIVIDUAL and FACILITY files.  Otherwise we'll end up with a lot of "duplicate" npi entries, each with slightly different information, including whether they're  `accepting` new patients.  We won't know which entry belongs to which plan.

Optional: Consider adding `issuer_id` (from PUF), as that would be useful for some of analytics.

**Comment by dportnoy** [comment]
https://github.com/demand-driven-open-data/ddod-intake/issues/56#issuecomment-213449685  Which state?  I'd love to get CA since BayesHack is there, but alas it's not included.  If you want to be on the conservative side, you can pick a state with fewest entries in PUF, like AK or SD.  When I did my last analysis a couple months ago, both these states had only one providers.json url.

**Comment by ghost** [comment]
Got it.  It will be interesting to see if the plans put leverage their own internal address information or just use the NPPES file...

I'll take a look at AK, I might also look at ME.

**Comment by ghost** [comment]
I am focussing on pulling files for ME.  Though it is only a few JSON files, the created tables are enormous.  Just for ME we are looking at probably a bit over 5GBs.  I am not sure what to do about that.  I guess I will just zip them up and post them.  Wow.

**Comment by dportnoy** [comment]
@BAP-Jeff, since the size is ending up to be so big anyway, what if we post a subset first.  Perhaps pick some arbitrary way to grab a subset that people can easily work with?

**Comment by ghost** [comment]
@dportnoy, okay I got Maine done.  It is posted to here:

[https://drive.google.com/file/d/0B9yZheZrBn54UFFBOGlLWmZVejQ/view?usp=sharing](url)

Compressed down to about 130mb.  Very little QA on this guy.  If we want to pull out something smaller let me know.  Not entirely obvious how to filter it...maybe by a couple PlanIDs.....

**Comment by dportnoy** [comment]
@BAP-Jeff nice!  Let's go with it.

**Comment by dportnoy** [comment]
Working on BayesImpact related posts now.  (@BAP-Jeff, @marks Could you help me summarize the latest and best links we have for each category.)
- DDOD Writeup: http://ddod.us/wiki/Challenge:_How_can_data_connect_individuals_with_the_health_providers_they_need
- BayesHack main website: http://bayeshack.org/health-and-human-services.html
- BayesHack GitHub to get devs started: https://github.com/bayesimpact/bayeshack-hhs-marketplace

---

**_Will finish tomorrow morning!_**

**Comment by marks** [comment]
@dportnoy all I've got for ya is https://healthdata.demo.socrata.com/view/xc22-8t66 for the latest version of the Formulary data @BAP-Jeff scraped. Happy to upload anything else too though

**Comment by dportnoy** [comment]
@marks, It would be great if you could load the updated data.  I'll proceed to publish individual links in the mean time, as well as data dictionaries.

**Comment by marks** [comment]
@dportnoy somehow missed Jeff's comment about ME. I'll see what I can do but please confirm that the Formulary file is as you'd expect it to be. Can definitely update title/description with whatever you'd like (perhaps a link back to the right place to see other resources)

**Comment by marks** [comment]
@BAP-Jeff / @dportnoy - as you all know, providers are split into 6 files this time. Think it's worth creating one or two combined files for easier analysis? May need to leave something off like fac type or language or concatenate arrays into a string. Just a suggestion for making use easier like it is for formulary. Regardless, working on uploading to enable viz and APIs for these 6 files

**Comment by ghost** [comment]
Good ideas.  I am out of pocket for most of the weekend.  Maybe we see if
anyone hits the APIs over the next couple weeks and then look to improve
things.

Will be interested to hear from @dportnoy what kind of interest/if any this
generates.  Clearly there are not many people using these files.

**Comment by marks** [comment]
@dportnoy the following files are ready

All US Forumulary file: https://healthdata.demo.socrata.com/view/xc22-8t66

Maine Provider Facility Type file: https://healthdata.demo.socrata.com/view/3juv-cnb4
Maine Provider Individual Specialty file: https://healthdata.demo.socrata.com/view/gz7p-vgqg
Maine Provider Individual Language file: https://healthdata.demo.socrata.com/view/fiik-zy3e

**Comment by dportnoy** [comment]
Update...  @BAP-Jeff, @marks, thank you again!  You guys were a huge help!  Couldn't have done it without you.  We still need to write up the activities at BayesHack, but there were 8 HHS teams at BayesHack, 4 of them specifically dealing with helping consumers find the right healthcare.  
![bayeshack - 8 hacker teams for hhs](https://cloud.githubusercontent.com/assets/2925801/14823613/c549e2c8-0b98-11e6-905d-0c3fa9929634.png)

**Comment by cornstein** [comment]
Anything useful/shareable come out of it? I would love to see.

**Comment by Jordanrau** [comment]
Ditto.

Jordan Rau | Senior Correspondent | Kaiser Health News<khn.org> | 202.654.1362 | @jordanrau | jrau@kff.orgmailto:jrau@kff.org
